src/main/java/sparksoniq/spark/iterator/flowr/ForClauseSparkIterator.java

```diff
@@ -195,4 +199,66 @@ public void close() {
         }
         return rdd;
     }
+
+
+    @Override
+    public Dataset<Row> getDataFrame(DynamicContext context) {
+        this._df = null;
+
+        if (this._child == null) {
+            // create initial RDD from expression
+            JavaRDD<Item> initialRdd = this.getNewRDDFromExpression(_expression);
+
+            // define a schema
+            String schemaString = _variableName;
+            List<StructField> fields = new ArrayList<>();
+            for (String fieldName : schemaString.split(" ")) {
+                StructField field = DataTypes.createStructField(fieldName, DataTypes.BinaryType, true);
+                fields.add(field);
+            }
+            StructType schema = DataTypes.createStructType(fields);
+
+            // convert initial RDD to row RDD
+            JavaRDD<Row> rowRDD = initialRdd.map(new ForClauseSerializeClosure());
+
+            // apply the schema to row RDD
+            this._df = SparkSessionManager.getInstance().getOrCreateSession().createDataFrame(rowRDD, schema);
+
+        } else {        //if it's not a start clause
+            if (_child.isDataFrame()) {
+                this._df = this._child.getDataFrame(context);
+
+                // TODO: deserialize the byte array dataframe
+                // line below and related closure is currently very experimental
+                Dataset<Row> ds = this._df.map(new ForClauseDeserializeClosure(), Encoders.bean(Row.class));
```

Comment:
I think that all the logics mapping an input Row to one or more output Rows (including the deserialization of the needed columns or reserialization of the newly added column) according to the for expression should be done inside a single new closure (so you probably do not need the deserialization closure). The output Row schema should probably be created here, it is not clear to me yet how the closure is informed of this output Row schema.